### PR TITLE
Create Initial Version of cuGraph Storage and Data Classes

### PR DESCRIPTION
**Pull request by alexbarghi-nv**
Adapts the TorchTensorGaasGraphDataProxy to PyG.  Makes some changes to PyG's typing in order to support ProxyTensors.  Adds CuGraphData and CuGraphStorage classes.

To-Do:

- [x] CuGraphData and CuGraph Storage
- [x] Modifications to other code in PyG to support CuGraphData/CuGraphStorage
- [x] Update to GaaS 
- [x] Fix data transfer to GPU
- [x] ~~Put notebook for end-to-end demo on new repository~~
- [x] Add GaaS example showing the end-to-end process as of now

### CODE REVIEW

**Comment by alexbarghi-nv** [comment]
Closes https://github.com/rapidsai/graph_dl/issues/11

File: torch_geometric/data/cugraph/cugraph_data.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+from torch_geometric.data.cugraph.cugraph_storage import CuGraphStorage
+
+from gaas_client.client import GaasClient
+from gaas_client.defaults import graph_id as DEFAULT_GRAPH_ID
+
+class CuGraphData(Data):
+    def __init__(self, gaas_client: GaasClient, graph_id: int=DEFAULT_GRAPH_ID, device=TorchDevice('cpu')):
+        super().__init__()
+
+        self.__dict__['_store'] = CuGraphStorage(gaas_client, graph_id, device=device)
```
Comment (rusty1s) [inline torch_geometric/data/cugraph/cugraph_data.py:12]:
```suggestion
        self._store = CuGraphStorage(gaas_client, graph_id, device=device)
```

File: torch_geometric/data/cugraph/cugraph_data.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        self.__dict__['device'] = device
```
Comment (rusty1s) [inline torch_geometric/data/cugraph/cugraph_data.py:13]:
```suggestion
        self.device = device
```

File: torch_geometric/nn/conv/message_passing.py [hunk 3]
```diff
@@ -184,7 +184,7 @@ def __set_size__(self, size: List[Optional[int]], dim: int, src: Tensor):
                  f'dimension {self.node_dim}, but expected size {the_size}.'))
 
     def __lift__(self, src, edge_index, dim):
-        if isinstance(edge_index, Tensor):
+        if isinstance(edge_index, Tensor) or isinstance(edge_index, ProxyTensor):
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:187]:
```suggestion
        if isinstance(edge_index, (Tensor, ProxyTensor)):
```

File: torch_geometric/nn/conv/message_passing.py [hunk 4]
```diff
@@ -148,7 +148,7 @@ def __init__(self, aggr: Optional[Union[str, List[str]]] = "add",
     def __check_input__(self, edge_index, size):
         the_size: List[Optional[int]] = [None, None]
 
-        if isinstance(edge_index, Tensor):
+        if isinstance(edge_index, Tensor) or isinstance(edge_index, ProxyTensor):
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:151]:
```suggestion
        if isinstance(edge_index, (Tensor, ProxyTensor)):
```

File: torch_geometric/nn/conv/message_passing.py [hunk 5]
```diff
@@ -220,7 +220,7 @@ def __collect__(self, args, edge_index, size, kwargs):
 
                 out[arg] = data
 
-        if isinstance(edge_index, Tensor):
+        if isinstance(edge_index, Tensor) or isinstance(edge_index, ProxyTensor):
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:223]:
```suggestion
        if isinstance(edge_index, (Tensor, ProxyTensor)):
```

File: torch_geometric/nn/conv/message_passing.py [hunk 6]
```diff
@@ -308,7 +308,7 @@ def propagate(self, edge_index: Adj, size: Size = None, **kwargs):
             out = self.update(out, **update_kwargs)
 
         # Otherwise, run both functions in separation.
-        elif isinstance(edge_index, Tensor) or not self.fuse:
+        elif isinstance(edge_index, Tensor) or isinstance(edge_index, ProxyTensor) or not self.fuse:
```
Comment (rusty1s) [inline torch_geometric/nn/conv/message_passing.py:311]:
```suggestion
        elif isinstance(edge_index, (Tensor, ProxyTensor)) or not self.fuse:
```

File: torch_geometric/typing.py [hunk 7]
```diff
@@ -26,14 +26,19 @@
 FeatureTensorType = Union[torch.Tensor, np.ndarray]
 
 # Types for message passing ###################################################
+class ProxyTensor(object):
```
Comment (rusty1s) [inline torch_geometric/typing.py:29]:
```suggestion
class ProxyTensor:
    pass
```

File: torch_geometric/typing.py [hunk 8]
```diff
@@ -26,14 +26,19 @@
...
+        return super()
 
-Adj = Union[Tensor, SparseTensor]
-OptTensor = Optional[Tensor]
-PairTensor = Tuple[Tensor, Tensor]
-OptPairTensor = Tuple[Tensor, Optional[Tensor]]
-PairOptTensor = Tuple[Optional[Tensor], Optional[Tensor]]
+AnyTensor = Union[Tensor, ProxyTensor]
+
+Adj = Union[AnyTensor, SparseTensor]
```
Comment (rusty1s) [inline torch_geometric/typing.py:35]: I am a bit worried that this blows up type hints in the documentation. I think we should leave this as it is and generally expect that `Tensor` also refers to `ProxyTensor`.

File: torch_geometric/data/cugraph/cugraph_data.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
+from torch import device as TorchDevice
+from torch_geometric.data import Data
+from torch_geometric.data.cugraph.cugraph_storage import CuGraphStorage
+
+from gaas_client.client import GaasClient
```
Comment (rusty1s) [inline torch_geometric/data/cugraph/cugraph_data.py:5]: I am not familiar with `Gass` but won't this lead to a circular dependency?

File: torch_geometric/data/cugraph/cugraph_data.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+class CuGraphData(Data):
+    def __init__(self, gaas_client: GaasClient, graph_id: int=DEFAULT_GRAPH_ID, device=TorchDevice('cpu')):
+        super().__init__()
+
+        self.__dict__['_store'] = CuGraphStorage(gaas_client, graph_id, device=device)
+        self.__dict__['device'] = device
+    
+    def to(self, to_device: TorchDevice) -> Data:
+        return CuGraphData(
+            self.__dict__['_store'].gaas_client,
```
Comment (rusty1s) [inline torch_geometric/data/cugraph/cugraph_data.py:17]:
```suggestion
            self.gaas_client,
```

File: torch_geometric/data/cugraph/cugraph_data.py [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+            self.__dict__['_store'].gaas_graph_id,
```
Comment (rusty1s) [inline torch_geometric/data/cugraph/cugraph_data.py:18]:
```suggestion
            self.gaas_graph_id,
```

File: torch_geometric/data/cugraph/cugraph_storage.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+from torch_geometric.typing import ProxyTensor
+from torch_geometric.data.storage import GlobalStorage
+
+from gaas_client.client import GaasClient
+
+EDGE_KEYS = ["_SRC_", "_DST_"]
+VERTEX_KEYS = ["_VERTEX_ID_"]
+
+
+class TorchTensorGaasGraphDataProxy(ProxyTensor):
```
Comment (rusty1s) [inline torch_geometric/data/cugraph/cugraph_storage.py:15]: This is interesting. I am wondering whether we should put the interface definition to `ProxyTensor` directly. The specific implementation is then properly better placed in `Gaas`?

Comment (alexbarghi-nv) [reply torch_geometric/data/cugraph/cugraph_storage.py:15]: We are currently debating this right now.  The question is whether this implementation of a ProxyTensor is general enough to include in GaaS.  I think we're leaning towards yes but haven't come to a firm decision just yet.

File: torch_geometric/data/cugraph/cugraph_data.py (hunk 9 above)
Comment (alexbarghi-nv) [reply torch_geometric/data/cugraph/cugraph_data.py:5]: GaaS isn't dependent on PyG, and CuGraphStorage isn't dependent on CuGraphData.

File: torch_geometric/typing.py (hunk 8 above)
Comment (alexbarghi-nv) [reply torch_geometric/typing.py:36]: I agree, but I don't think we want to subclass torch.Tensor as that comes with some additional requirements.

Comment (alexbarghi-nv) [reply torch_geometric/typing.py:36]: I'm not sure what the workaround is other than subclassing torch.Tensor, but I will look into an alternative solution.

Comment (rusty1s) [reply torch_geometric/typing.py:36]: I don't think we need a workaround. Type hints do not do any type checking, they are just hints for the user. As such, we can just treat `Tensor` and `ProxyTensor` as the same.

File: torch_geometric/data/cugraph/cugraph_data.py (hunk 9 above)
Comment (rusty1s) [reply torch_geometric/data/cugraph/cugraph_data.py:5]: Then I might misunderstanding the overall roadmap here. Is the plan to have `GaasClient` as a dependency inside PyG?

File: torch_geometric/data/cugraph/cugraph_storage.py (hunk 12 above)
Comment (rusty1s) [reply torch_geometric/data/cugraph/cugraph_storage.py:15]: I think it is definitely of interest in PyG as well in order to connect to different backends. This is similar to our efforts of bringing a `FeatureStore` abstraction layer to PyG, see: https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/data/feature_store.py

Eventually, these concepts can be unified. Pinging @mananshah99 for additional thoughts!

File: torch_geometric/data/cugraph/cugraph_data.py (hunk 9 above)
Comment (alexbarghi-nv) [reply torch_geometric/data/cugraph/cugraph_data.py:5]: optional dependency, eventually.

**Pull request merged**